test/benchmarks/src/main/scala/scala/collection/mutable/ArrayDequeBenchmark.scala

```diff
@@ -0,0 +1,147 @@
+package scala.collection
+package mutable
+
+import java.util.concurrent.TimeUnit
+
+import scala.util.Random
+
+import org.openjdk.jmh.annotations._
+import org.openjdk.jmh.infra.Blackhole
+
+@BenchmarkMode(Array(Mode.AverageTime))
+@Fork(2)
+@Threads(1)
+@Warmup(iterations = 10)
+@Measurement(iterations = 10)
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+@State(Scope.Benchmark)
+class ArrayDequeBenchmark {
+
+  @Param(Array("0", "1000"))
+  var length: Int = _
+
+  var arr: ArrayDeque[Int] = _
+  var fraction: Int = _
+  var near: Int = _
+  var chunk: Int = _
+  var tail: Int = _
+  var tailChunk: Int = _
+  var nearTail: Int = _
+  var nearTailChunk: Int = _
+
+  @Param(Array("1000"))
+  var range: Int = _
+
+  var rangeArr: Array[Int] = _
+
+  @Setup(Level.Trial)
+  def initRange: Unit = {
+     rangeArr = (1 to range.toInt).toArray
+  }
+
+  @Setup(Level.Invocation)
+  def initArray: Unit = {
+    arr = ArrayDeque.range(0, length)
+    fraction = 20
```

Comment:
I recall trying to use `Level.Trial` and `Level.Iteration`, but there were issues since ArrayDeque is mutable.  I'll try again.

Reply:
I tried changing the `Setup` to both again, but still get an exception thrown:

    java.lang.IndexOutOfBoundsException: 929 is out of bounds (min 0, max 928)

Not sure why it's doing that.

Reply:
The ArrayDeque has to w recreated each invocation, I was talking about the other constants

Reply:
I see.  So make it like this?

```scala
  @Setup(Level.Trial)
  def setupTrial: Unit = {
    rangeArr = (1 to range.toInt).toArray
    fraction = 20
    near = length / fraction
    chunk = Math.min(length, fraction)
    tail = Math.max(0, length - 1)
    tailChunk = Math.max(0, tail - chunk)
    nearTail = Math.max(0, tail - near)
    nearTailChunk = Math.max(0, nearTail - chunk)
  }

  @Setup(Level.Invocation)
  def setupInvocation: Unit = {
    arr = ArrayDeque.range(0, length)
  }
```

I had considered it, but I liked the names `initArray` and `initRange`.

Reply:
OK.  So in general using Level.Invocation can make JMH less precise.  One way to avoid that would be to benchmark a remove and an insert at the same time. But your code works well enough so seems fine.

Reply:
JMH is supposed to ensure the benchmark is isolated from any setup and teardown.  We have to do it here, because ArrayDeque is mutable.  

Also, I'm sure allocating the ArrayDeque is a massive computational part of each invocation setup, and the six arithmetic calculations are a drop in the bucket.

Reply:
>JMH is supposed to ensure the benchmark is isolated from any setup and teardown.

http://javadox.com/org.openjdk.jmh/jmh-core/1.7/org/openjdk/jmh/annotations/Level.html#Invocation